File: 11572 - Unique Snowflakes/11572 - Unique Snowflakes.cpp
#include <bits/stdc++.h>
using namespace std;
int main(){
    map <int, bool> dic;
    deque <int> q;
    int cases,res, cant, snowflake;
    scanf("%d",&cases);
    while (cases--){
        scanf("%d",&cant);
        dic.clear();
        q.clear();
        res = 0;
        while (cant--){
            scanf("%d",&snowflake);
            if (dic.find(snowflake) == dic.end() || dic[snowflake] == false){
                dic[snowflake] = true;
                q.push_back(snowflake);
            }
            else{
                res = max(res, (int)q.size());
                q.push_back(snowflake);
                while (dic[snowflake]){
                    dic[q.front()] = false;
                    q.pop_front();
                }
                dic[snowflake] = true;
            }
            res = max(res, (int)q.size());
        }
        printf("%d\n", res);
    }
    return 0;
}

File: 11572 - Unique Snowflakes/11572 - Unique Snowflakes.py
from sys import stdin as s
from collections import deque
def main():
    cases = int(s.readline().strip())
    while cases:
        cant = int(s.readline().strip())
        dic = {}
        queue = deque()
        res = 0
        while cant:
            snowflake = s.readline().strip()
            if snowflake not in dic or dic[snowflake] == False:
                dic[snowflake] = True
                queue.append(snowflake)
            else:
                res = max(res, len(queue))
                queue.append(snowflake)
                while dic[snowflake]:
                    value = queue.popleft()
                    dic[value] = False
                dic[snowflake] = True
            cant -= 1
        res = max(res, len(queue))
        print(res)
        cases -= 1
main()